### PR TITLE
feat(client): warn when non-.cdn endpoints are used

### PR DESCRIPTION
**Pull request by lihbr**
## Types of changes

- [ ] Chore (a non-breaking change which is related to package maintenance)
- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Resolves: #284

## Checklist:

- [ ] My change requires an update to the official documentation.
- [x] All [TSDoc](https://tsdoc.org) comments are up-to-date and new ones have been added where necessary.
- [x] All new and existing tests are passing.

### CODE REVIEW

File: messages/endpoint-must-use-cdn.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# `endpoint` must use CDN
+
+`@prismicio/client` uses either a repository name or a repository endpoint to query content from Prismic.
```
Comment (angeloashmore) [inline messages/endpoint-must-use-cdn.md:3]: Could we change "repository endpoint" to "Prismic Rest API V2 repository endpoint", just so it's abundantly clear what is accepted?

> `@prismicio/client` uses either a Prismic repository name or a Prismic Rest API V2 repository endpoint to query content from Prismic.

File: messages/endpoint-must-use-cdn.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+The repository name or repository endpoint must be provided when creating a `@prismicio/client` like the following:
+
+```typescript
+import * as prismic from "@prismicio/client";
+
+// Using the repository name
+const client = prismic.createClient("example")
+
+// Using the repository endpoint
+const client = prismic.createClient("https://example.cdn.prismic.io/api/v2")
```
Comment (angeloashmore) [inline messages/endpoint-must-use-cdn.md:14]: Going forward, @samlfair recommends we use `example-prismic-repo` when needing a placeholder repository name. The repository is real and contains some example content.

File: messages/endpoint-must-use-cdn.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+```
+
+When creating a `@prismicio/client` with a repository endpoint, endpoint's subdomain must feature the `.cdn` suffix.
```
Comment (angeloashmore) [inline messages/endpoint-must-use-cdn.md:17]: Missing a word:

> When creating a `@prismicio/client` with a repository endpoint, **the** endpoint's subdomain must feature the `.cdn` suffix.

File: messages/endpoint-must-use-cdn.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import * as prismic from "@prismicio/client";
+
+// ✅ Correct
+const client = prismic.createClient("https://example.cdn.prismic.io/api/v2")
+
+// ❌ Incorrect
+const client = prismic.createClient("https://example.prismic.io/api/v2")
+```
+
+Not using the `.cdn` version of your repository endpoint can have unexpected side-effects and cause performance issues when query Prismic.
```
Comment (angeloashmore) [inline messages/endpoint-must-use-cdn.md:29]: "query" -> "querying"

> Not using the `.cdn` version of your repository endpoint can have unexpected side-effects and cause performance issues when **querying** Prismic.

File: src/createClient.ts [hunk 5]
```diff
@@ -501,15 +503,37 @@ export class Client<TDocuments extends PrismicDocument = PrismicDocument> {
...
-				throw new PrismicError(
-					"@prismicio/client only supports Prismic Rest API V2. Please provide only the repository name to the first createClient() parameter or use the getRepositoryEndpoint() helper to generate a valid Rest API V2 endpoint URL.",
-					undefined,
-					undefined,
-				);
+			if (process.env.NODE_ENV === "development") {
+				/**
+				 * Matches non-API v2 `.prismic.io` endpoints
+				 * {@link https://regex101.com/r/xRsavu/1}
+				 */
```
Comment (angeloashmore) [inline src/createClient.ts:510]: Style: I typically use `/* ... */` comments only for TSDoc (i.e. ones that should be public) and `//` for internal comments. Following that convention makes it easy to put TSDocs right next to internal comments, like this:

```typescript
/* A public comment. */
// TODO: A private comment
function foo() {
  return "bar"
}
```

`foo()` will retain the TSDoc even with the `//` comment between the documentation and the function.

File: test/client.test.ts [hunk 6]
```diff
@@ -80,6 +80,75 @@ it("constructor throws if an invalid repository endpoint is provided", () => {
...
+
+	expect(() => {
+		prismic.createClient("https://qwerty.cdn.prismic.io/api/v2", { fetch });
+	}, "A valid prismic.io V2 endpoint does not throw").not.toThrow();
+
+	expect(() => {
+		prismic.createClient(prismic.getRepositoryEndpoint("qwerty"), { fetch });
+	}, "An endpoint created with getRepositoryEndpoint does not throw").not.toThrow();
+
+	process.env.NODE_ENV === originalNodeEnv;
```
Comment (angeloashmore) [inline test/client.test.ts:108]: Oops, there was a typo in the old test. This line should be:

```typescript
process.env.NODE_ENV = originalNodeEnv
```

File: test/client.test.ts [hunk 7]
```diff
@@ -80,6 +80,75 @@ it("constructor throws if an invalid repository endpoint is provided", () => {
...
+
+	prismic.createClient(prismic.getRepositoryEndpoint("qwerty"), { fetch });
+	expect(
+		consoleWarnSpy,
+		"An endpoint created with getRepositoryEndpoint does not warn",
+	).not.toHaveBeenCalledWith(expect.stringMatching(/endpoint-must-use-cdn/i));
+
+	consoleWarnSpy.mockRestore();
+
+	process.env.NODE_ENV === originalNodeEnv;
```
Comment (angeloashmore) [inline test/client.test.ts:149]: Same comment as above:

```typescript
process.env.NODE_ENV = originalNodeEnv
```

**Review by angeloashmore — CHANGES_REQUESTED** [review]
Looks great! Most of the comments I left are about style or rephrasing the messages. The technical implementation looks perfect. 👍

File: messages/endpoint-must-use-cdn.md (hunk 4 above)
Comment (lihbr) [reply messages/endpoint-must-use-cdn.md:29]: I don't know how to write in English anymore 🤦‍♀️

File: test/client.test.ts (hunk 6 above)
Comment (lihbr) [reply test/client.test.ts:108]: Nice catch!

**Pull request merged**